File: 0077-Combinations/lc0077.cc
// C++ solution.

class Solution {
public:
  vector<vector<int>> combine(int n, int k) {
    vector<vector<int>> ans;
    vector<int> res(k, 0);
    dfs(ans, res, 1, n, k);
    return ans;
  }

  void dfs(vector<vector<int>>& ans, vector<int>& res, int start, const int n, const int k) {
    if (k == 0) {
      ans.push_back(res);
    } else {
      for (int i = start; i <= (n-k+1); i++) {
        res[res.size()-k] = i;
        dfs(ans, res, i+1, n, k-1);
      }
    }
  }
};

File: 0077-Combinations/lc0077.rs
// Rust solution.

impl Solution {
    pub fn combine(n: i32, k: i32) -> Vec<Vec<i32>> {
        if k == 0 {
            return vec![vec![]];
        }

        let mut ans: Vec<Vec<i32>> = Vec::new();
        for i in k..=n {
            // previous combine result will be moved here
            for mut pre in Self::combine(i-1, k-1) {
                pre.push(i);
                ans.push(pre);
            }
        }
        ans
    }
}
